File: src/main/java/com/demonwav/mcdev/platform/mcp/McpModule.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mcp;

import com.demonwav.mcdev.facet.MinecraftFacet;
import com.demonwav.mcdev.platform.AbstractModule;
import com.demonwav.mcdev.platform.PlatformType;
import com.demonwav.mcdev.platform.mcp.srg.SrgManager;
import com.google.common.collect.Sets;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiMethod;
import java.util.Set;
import javax.swing.Icon;
import org.jetbrains.annotations.NotNull;

public class McpModule extends AbstractModule {

    private final McpModuleSettings settings;
    private final Set<VirtualFile> accessTransformers = Sets.newHashSet();
    private SrgManager srgManager = new SrgManager();

    public McpModule(@NotNull MinecraftFacet facet) {
        super(facet);

        this.settings = McpModuleSettings.getInstance(module);
        srgManager.parse(getSettings().getMappingFiles());
    }

    @Override
    public McpModuleType getModuleType() {
        return McpModuleType.INSTANCE;
    }

    @Override
    public PlatformType getType() {
        return PlatformType.MCP;
    }

    @Override
    public Icon getIcon() {
        return null;
    }

    @Override
    public String writeErrorMessageForEventParameter(PsiClass eventClass, PsiMethod method) {
        return "";
    }

    public McpModuleSettings.State getSettings() {
        return settings.getState();
    }

    public SrgManager getSrgManager() {
        return srgManager;
    }

    public void updateSettings(McpModuleSettings.State data) {
        this.settings.loadState(data);
        srgManager.parse(data.getMappingFiles());
    }

    public Set<VirtualFile> getAccessTransformers() {
        return accessTransformers;
    }

    public void addAccessTransformerFile(@NotNull VirtualFile file) {
        accessTransformers.add(file);
    }
}

File: src/main/kotlin/com/demonwav/mcdev/platform/mcp/McpModule.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.platform.mcp

import com.demonwav.mcdev.facet.MinecraftFacet
import com.demonwav.mcdev.i18n.I18nFileListener
import com.demonwav.mcdev.platform.AbstractModule
import com.demonwav.mcdev.platform.PlatformType
import com.demonwav.mcdev.platform.mcp.srg.SrgManager
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiMethod
import com.intellij.util.messages.MessageBusConnection
import javax.swing.Icon

class McpModule(facet: MinecraftFacet) : AbstractModule(facet) {

    private lateinit var connection: MessageBusConnection

    private val settings: McpModuleSettings = McpModuleSettings.getInstance(module)
    val accessTransformers = mutableSetOf<VirtualFile>()

    var srgManager: SrgManager? = null
        private set

    override fun init() {
        initSrg()
        connection = project.messageBus.connect()
        connection.subscribe(VirtualFileManager.VFS_CHANGES, I18nFileListener)
    }

    private fun initSrg() {
        val settings = getSettings()
        val file = settings.mappingFile ?: return
        val srgType = settings.srgType ?: return

        srgManager = SrgManager.getInstance(file, srgType)
        srgManager?.parse()
    }

    override val moduleType = McpModuleType
    override val type = PlatformType.MCP
    override val icon: Icon? = null

    override fun writeErrorMessageForEventParameter(eventClass: PsiClass, method: PsiMethod) = ""

    private fun getSettings() = settings.state

    fun updateSettings(data: McpModuleSettings.State) {
        this.settings.loadState(data)
        val mappingFile = data.mappingFile ?: return
        val srgType = data.srgType ?: return

        srgManager = SrgManager.getInstance(mappingFile, srgType)
        srgManager?.parse()
    }

    fun addAccessTransformerFile(file: VirtualFile) {
        accessTransformers.add(file)
    }

    override fun dispose() {
        super.dispose()

        connection.disconnect()
        accessTransformers.clear()
        srgManager = null
    }
}
